docs/sql-programming-guide.md

```diff
@@ -1002,6 +1002,21 @@ Configuration of Parquet can be done using the `setConf` method on `SparkSession
     </p>
   </td>
 </tr>
+<tr>
+  <td><code>spark.sql.parquet.writeLegacyFormat</code></td>
+  <td>false</td>
+  <td>
+    This configuration indicates whether we should use legacy Parquet format adopted by Spark 1.4
+    and prior versions or the standard format defined in parquet-format specification to write
+    Parquet files. This is not only related to compatibility with old Spark ones, but also other
+    systems like Hive, Impala, Presto, etc. This is especially important for decimals. If this
+    configuration is not enabled, decimals will be written in int-based format in Spark 1.5 and
+    above, other systems that only support legacy decimal format (fixed length byte array) will not
+    be able to read what Spark has written. Note other systems may have added support for the
+    standard format in more recent versions, which will make this configuration unnecessary. Please
```

Comment:
Yeah, I think Hive and Impala also use newer Parquet versions/format. Isn't it sufficient to say older versions of Spark (<= 1.4) and older versions of Hive, Impala (do we know which?) use older Parquet formats and this enables writing it that way?

Reply:
I haven't checked closely but I think Hive still uses binary for decimals (https://github.com/apache/hive/blob/ae008b79b5d52ed6a38875b73025a505725828eb/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/write/DataWritableWriter.java#L503-L541). Given my past investigation, thing is, Parquet supports both ways to write out (https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#decimal) IIRC. They deprecated timestamp based on int 96 (https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L782) but not decimals.

Reply:
This is another issue since we call the option something "legacy" which isn't actually legacy in Parquet's decimal side.

Reply:
Hive and Impala do NOT support new Parquet format yet.

* [HIVE-19069](https://jira.apache.org/jira/browse/HIVE-19069): Hive can't read int32 and int64 Parquet decimal. This issue is not resolved yet. This is consistent with source code check by @HyukjinKwon 
* [IMPALA-5542](https://issues.apache.org/jira/browse/IMPALA-5542): Impala cannot scan Parquet decimal stored as int64_t/int32_t. This is resolved, however targeted to Impala 3.1.0, which is a version not released yet. The latest release is 3.0.0 (https://impala.apache.org/downloads.html).

Presto began to support new Parquet format since 0.182.

* [issues/7533](https://github.com/prestodb/presto/issues/7533): Improve decimal type support in the new Parquet reader.  This patch is included in [0.182](https://prestodb.io/docs/current/release/release-0.182.html). Blow is the excerpt: 

> Fix reading decimal values in the optimized Parquet reader when they are backed by the int32 or int64 types.

Reply:
It sounds like it isn't quite a legacy format, but one still used by Hive and even considered valid if not current by Parquet? This part I am not sure of, but basing it on Hyukjin's comment above. 

I suggest a somewhat shorter text like this, what do you think? its length would be more suitable as a config doc below.

If `true`, then decimal values will be written in Apache Parquet's fixed-length byte array format. This is used by Spark 1.4 and earlier, and systems like Apache Hive and Apache Impala. If `false`, decimals will be written using the newer int format in Parquet. If Parquet output is intended for use with systems that do not support this newer format, set to `true`.

Reply:
If we must call it "legacy", I'd think of it legacy implementation in Spark side, rather than legacy format in Parquet side. 
As comment in [SPARK-20297](https://issues.apache.org/jira/browse/SPARK-20297?focusedCommentId=15975559&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15975559)
> The standard doesn't say that smaller decimals have to be stored in int32/int64, it just is an option for subset of decimal types. int32 and int64 are valid representations for a subset of decimal types. fixed_len_byte_array and binary are a valid representation of any decimal type.
> 
>The int32/int64 options were present in the original version of the decimal spec, they just weren't widely implemented. So its not a new/old version thing, it was just an alternative representation that many systems didn't implement.

Anyway, it really leads to confusion.

Really appreciate your suggestion @srowen to make the doc shorter, the doc you suggested is more concise and to the point.

One more thing I want to discuss. After investigating the usage of this option,  I found this option is not only related to decimals, but also complex types (Array, Map), see below source code. Should we mention this in the doc?

https://github.com/apache/spark/blob/473d0d862de54ec1c7a8f0354fa5e06f3d66e455/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala#L450-L458

Reply:
Let's make it short and get rid of all other things orthogonal with the issue itself (I think the issue is specific to decimals). For instance, we could say:

If `true`, it writes Parquet file in a way of Spark 1.4 and earlier, for instance, decimal values will be written in Apache Parquet's fixed-length byte array format, which other systems such as Apache Hive and Apache Impala use. If `false`, the newer format in Parquet will be used, for instance, decimals will be written based on int. If Parquet output is intended for use with systems that do not support this newer format, set to `true`.

Please feel free to change words as what you think is righter